### ISSUE TITLE
Retain comments inside type annotations during Pallene to Lua translation.

### ISSUE THREAD

**Issue by itssamuelrowe** [pallene-lang/pallene#273]
As of now, comments before the actual type annotation begins are retained. See the snippet
below:
```
local i : -- This comment is retained.
{
   -- This comment will be erased.
   integer
} = { 1, 2, 3 }
```

This is possible because we have access to enough offsets to retain such comments. However, comments that appear within type annotations are erased. One way to solve this is to create a small parser using lpeglabel that detects comments inside type annotations. An example of such a parser can be found in `C.lua` where we employ this technique to detect C comments.

**Comment by hugomg** [comment]
Here is a suggestion of how this could be done:

1. Get rid of all the _loc fields from the AST. Instead of adding the start and end locations of type annotations to the AST, we could have the parser return a list of type annotation regions in addition to returning the AST. This would let us get rid of the _loc fields that are cluttering the AST and it would also allow us to get rid of most of the recursive functions in the translator.lua.

2. One way to do this would be to add a pair of functions to the parser called `begin_type_annotation` and `end_type_annotation`, to mark the begin and end of the region that we will want to erase in the Lua translator. Similarly to the `begin_loop` and `end_loop` functions.

3. Modify the Lexer:next method so that it does not skip comments. It will now be the responsibility of the parser to skip those comments, presumably in the _advance method. You may need to fix some test cases in the lexer spec to include the `"COMMENT"` outputs.

4. Carefully ensure that comments are left out of the regions of code that are marked to be erased. For example, if we have a piece of code like `local i: { --[[this is a comment]] integer }` then it should mark the `: { ` region and the ` integer }` region as containing type annotations that need to be erased.

We could do this is a single PR or we couldd split part 1+2 in one PR and 3+4 in another PR.

**Comment by hugomg** [comment]
Implemented in #306
